### PR TITLE
Clean stale object files from `SwiftCompile` task

### PR DESCRIPTION
**Pull request by lacasseio**
### Context
See https://github.com/gradle/gradle-native/issues/94.

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and [CI build status](https://builds.gradle.org/project.html?projectId=Gradle_Check&tab=projectOverview&branch_Gradle_Check=dl%2Fswift-stale-object-coverage)
- [x] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs
- [x] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy [hunk 1]
```diff
@@ -104,4 +107,68 @@ class SwiftIncrementalCompileIntegrationTest extends AbstractInstalledToolChainI
...
+        and:
+        succeeds "assemble"
+        app.applyChangesToProject(testDirectory)
+
+        expect:
+        succeeds "assemble"
+        result.assertTasksExecuted(":compileDebugSwift", ":linkDebug", ":installDebug", ":assemble")
+        result.assertTasksNotSkipped(":compileDebugSwift", ":linkDebug", ":installDebug", ":assemble")
+
+        files("build/obj/main")*.name as Set == app.alternate.expectedIntermediateFilenames
```
Comment (big-guy) [inline subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy]: I think you can use...

`file("build/obj/main").assertHasDescendants(app.alternate.expectedIntermediateFilenames)`

File: subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/SwiftCompile.java [hunk 2]
```diff
@@ -78,4 +80,18 @@ public void setModuleName(String moduleName) {
...
         this.moduleName.set(moduleName);
     }
+
+    @Override
+    public void compile(IncrementalTaskInputs inputs) {
+        SimpleStaleClassCleaner cleaner = new SimpleStaleClassCleaner(getOutputs());
+        cleaner.setDestinationDir(getObjectFileDir());
+        cleaner.execute();
+
+        if (getSource().isEmpty()) {
```
Comment (big-guy) [inline subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/SwiftCompile.java]: I think this might be unnecessary if `getSource` is annotated with `@SkipWhenEmpty`. I think we delete the outputs when all the sources are removed.

File: subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/SwiftCompile.java [hunk 3]
```diff
@@ -78,4 +80,18 @@ public void setModuleName(String moduleName) {
     public void setModuleName(Provider<String> moduleName) {
         this.moduleName.set(moduleName);
     }
+
+    @Override
+    public void compile(IncrementalTaskInputs inputs) {
+        SimpleStaleClassCleaner cleaner = new SimpleStaleClassCleaner(getOutputs());
```
Comment (big-guy) [inline subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/SwiftCompile.java]: Isn't this just a fancy way of deleting all of the outputs every time?

There's this TODO:
https://github.com/gradle/gradle/blob/85c23b8dc5a1edfe0e957c041f24d71179bbe792/subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftPlatformToolProvider.java#L73-L74

Is this related? Shouldn't we already be deleting stale outputs?

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy [hunk 4]
```diff
@@ -104,4 +107,68 @@ class SwiftIncrementalCompileIntegrationTest extends AbstractInstalledToolChainI
...
+        expect:
+        succeeds "assemble"
+        result.assertTasksExecuted(":compileDebugSwift", ":linkDebug", ":assemble")
+        result.assertTasksNotSkipped(":compileDebugSwift", ":linkDebug", ":assemble")
+
+
+        files("build/obj/main")*.name as Set == lib.alternate.expectedIntermediateFilenames
+        sharedLibrary("build/lib/main/debug/Hello").assertExists()
+    }
+
```
Comment (big-guy) [inline subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy]: Should we have a test for a project where the sources are completely removed?

File: subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/SwiftCompile.java (hunk 3 above)
Comment (adammurdoch) [reply subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/SwiftCompile.java]: That `OutputCleaningCompiler` won't actually do anything, because there's nothing upstream of it calculating which object files need to be removed, which is what it's expecting. It just shouldn't be there at all.

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy (hunk 4 above)
Comment (adammurdoch) [reply subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy]: Yes we should. We should make sure all the tasks have the correct outcomes and the outputs are gone in this instance.

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy (hunk 1 above)
Comment (adammurdoch) [reply subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy]: even if you can't, the `files()` method looks like something that belongs on `TestFile` rather than sprinkled through a bunch of test classes

Comment (lacasseio) [reply subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy]: I did some modification to the fixture so we can use `TestFile#assertHasDescendants()`.

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy (hunk 4 above)
Comment (lacasseio) [reply subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy]: This will be done as a separate PR addressing stale link output.

File: subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/SwiftCompile.java (hunk 3 above)
Comment (lacasseio) [reply subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/SwiftCompile.java]: Adam is right on this, we shortcut the incremental compiler creation by returning the normal, non-incremental, compiler. The issue is `IncrementalNativeCompiler` will do the wrong thing as it's opinionated toward `#include` processing. This will be addressed when we roll incremental support for Swift. Right now, we just want the stale object file to be cleaned up.

File: subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/SwiftCompile.java (hunk 2 above)
Comment (lacasseio) [reply subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/SwiftCompile.java]: I remove that part and didn't add `@SkipWhenEmpty` as removing all source will be tested in another PR.

**Comment by adammurdoch** [comment]
It's fine to add the additional test in another PR, once all of this stuff is merged.
